### PR TITLE
Refactor ACL persistent storage and events

### LINKED ISSUE #14443: Clean up access control cluster/module initialization
In the cluster initialization, make it responsible for initializing the system module, populating the data (from flash storage), reporting unrecoverable errors to system startup, etc.

This probably needs to (or at least should) be done before #13867.

### LINKED ISSUE #14451: Improve access control cluster persistence/storage
Currently (earlier issues and PRs) we have simple persistence using custom/duplicated TLV code at the system module layer with light metadata.

We'd like to use more of the generated code of course. This means moving persistence to the cluster layer.

There are other improvements that can be made (e.g. only serializing entries that are in use). Basically, this code just needs more attention over the "minimal viable code" that currently exists.

### PR DESCRIPTION
**Pull request by mlepage-google**
**PLEASE NOTE I would like the following commit message used when this is squashed and merged**

Other notes:
- basically ready for review, but I'm still checking a couple little things
- I know we have API duplication, I'll remove the older APIs in a subsequent PR
- Basically I'll change it all to fabric-based indexing, but I don't want this PR to be larger than it has to be

#### Commit Message
Refactor ACL persistent storage and events

After these changes, the breakdown of responsibilities is as follows:
- System module is the API for access control (entries, check) used by IM, cluster, etc.
  - Only responsible for live ACL entries in memory, not responsible for persistent storage etc.
- System module delegate is responsible for implementation (e.g. ExampleAccessControlDelegate)
- AclStorage is responsible for all ACL entry translation between system API, cluster generated code, and TLV
  - Also responsible for all loading/saving of ACL entries in persistent storage
- Cluster is responsible for implementing ACL attribute changes (using system module API)
  - Also responsible for creating events when ACL entries are changed
  - Also responsible for extension attribute (which only exists in cluster, not in system module)
- Server is responsible for initializing these modules in the correct order
  - Including deleting ACL entries when fabric is removed

Details:
- Add fabric-centric APIs to system module (older fabric-optional APIs will be removed later)
- Add API to determine entry count per fabric
- Add API to determine if an entry is using the default delegate
- Add API for dumping an entry to log (disabled, because only used during development)
- Add notification listener API to system module for when ACL entries are changed
- Add subject descriptor to system module APIs so ACL entry changes can (optionally) be attributed
- Add AclStorage module to manage encoding/decoding/storage of ACL entries
- Remove encoding/decoding/storage responsibilities from system module and cluster now that AclStorage exists
- Add ACL entry encoder and decoder helper classes which can stage/unstage (transcode) and TLV encode/decode
- Add accurate calculation of required storage buffer size for TLV encoded entry
- Refactor server initialization order so system module is initialized first, then AclStorage, then cluster
- Use fabric-specific persistent storage keys for ACL entries
- AclStorage loads entries from persistent storage to populate the system module during server initialization
- AclStorage installs a listener in the system module to keep persistent storage in sync with in-memory storage
- Cluster installs a listener in the system module to notify events when ACL entries are changed
- Delete ACL entry event now occurs after (was before) actual ACL entry deletion
- System module delegate example implementation now makes (best effort) attempt to preserve deleted entries
- All entry changes (from cluster, remove fabric, etc.) now performed in most efficient order (typically LIFO)
- All entry persistent storage changes (from AclStorage) now only touches affected persistent storage keys
- Operational credentials cluster now creates intial ACL for fabric after (was before) adopting the fabric

Fixes #14451 --> Improve access control cluster persistence/storage
Fixes #14443 --> Clean up access control cluster/module initialization
Fixes #14326 --> Handle access control events from non-cluster operations

Partially addresses #14455 --> Clean up access control capabilities APIs
Partially addresses #14450 --> Access control performance improvement (flash, RAM, speed, etc.)

#### Testing
- Local testing (Linux x64, using chip-all-clusters-app and chip-tool)
- CI

### CODE REVIEW

**Review by andy31415 — DISMISSED** [review]
Requesting changes to not accidentally merge this.

Please remove only once/if it is actually ready

**Comment by mlepage-google** [comment]
Thanks, this PR is **not ready for review yet**, but it's progressed far enough to get a **flash size estimate** while I work on other things.

**Comment by mlepage-google** [comment]
This PR is **ready for review**. I might still tweak one or two things I am double checking but it should be in good shape, and I have a bit more to do in subsequent PRs but it shouldn't hold up this one. Please note that **I would like the giant commit message in the PR description when it is squashed and merged**, and not just whatever github does by default.

File: src/access/examples/ExampleAccessControlDelegate.cpp
Comment (tcarmelveilleux) [inline src/access/examples/ExampleAccessControlDelegate.cpp:728]: Not clear what this does

File: src/app/server/AclStorage.cpp
Comment (tcarmelveilleux) [inline src/app/server/AclStorage.cpp:190]:
```suggestion
        ReturnErrorCodeIf((from.nodeId & ~kMaskPAKEKeyId) != 0, CHIP_ERROR_INVALID_ARGUMENT);
```

Comment (mlepage-google) [reply src/app/server/AclStorage.cpp]: Added the `!= 0` to all the places using those flags.

**Comment by mlepage-google** [comment]
@tcarmelveilleux requests:
`Do not reach into Server for the PersistentStorage. We inject the AccessControl::Delegate to the Server SPECIFICALLY so that the init of it injects the persistent storage delegate. It has to be kept this way.`
`Initialize the items to inject PersistentStorageDelegate, rather than reaching in for auto & storage = Server::GetInstance().GetPersistentStorage();`

I will change it so we inject the persistent storage into AclStorage, and update docs.

File: src/app/server/Server.cpp
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:151]: Did this get moved down because this way we don't queue events as we read the fabrics from storage?  If so, the ordering dependency should be documented here.  If not, why was it moved?

File: src/app/server/AclStorage.h
Comment (bzbarsky-apple) [inline src/app/server/AclStorage.h:80]: Why is mStagingEntry a member?  It's just used temporarily; it could go on the stack...

Or is the idea that this whole class normally just lives on the stack?  Worth documenting that consideration, if so.

File: src/app/server/AclStorage.h
Comment (bzbarsky-apple) [inline src/app/server/AclStorage.h:99]: How does the fabric index get used here?

File: src/app/server/AclStorage.cpp
Comment (bzbarsky-apple) [inline src/app/server/AclStorage.cpp:37]: Is there a reason not to use `TLV::EstimateStructOverhead` here?  I guess it might be a pain for the lists... Worth at least a comment about why it's not being used.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/access/AccessControl.cpp
Comment (mlepage-google) [reply src/access/AccessControl.cpp:405]: This is only here for ad hoc use during development, in case the developer needs to see what an entry looks like. It's not normally used. I changed it to detail anyways.

File: src/app/server/AclStorage.cpp
Comment (mlepage-google) [reply src/app/server/AclStorage.cpp:37]: I'm still going to have to come back and revisit that last TODO (in another PR, I'd like to get this one in) so I'll add any relevant comments about these functions when I do that (I made a note of it).

File: src/access/AccessControl.h
Comment (bzbarsky-apple) [reply src/access/AccessControl.h:460]: I don't undertand this part.  We are creating an entry.  We pass in the entry to create.  It's always present (though we do return early if it's not valid).  In what sense is it best-effort or optional?

File: src/lib/support/DefaultStorageKeyAllocator.h
Comment (bzbarsky-apple) [reply src/lib/support/DefaultStorageKeyAllocator.h]: Yeah, I don't know that any of our CI _runs_ the things with minimal libcs, and the error is a runtime one, not a compile-time one (mis-handling the actual formatting).

File: src/access/AccessControl.h
Comment (mlepage-google) [reply src/access/AccessControl.h:460]: Sorry I was thinking about the listener entry, not the create entry. Yes this could use cleanup (or documentation) because of this detail. I do have to (future PR) remove the old fabric-index-optional APIs and I will tidy this up to. Short answer is, now that fabric index based indexing is required (not optional), we use that explicit fabric index and ignore what's in the entry. Next PR I'll formalize it all more.

File: src/access/AccessControl.h
Comment (tcarmelveilleux) [inline src/access/AccessControl.h:317]: Is it set to nullptr if not available?

Comment (mlepage-google) [reply src/access/AccessControl.h:317]: Yes it's a pointer.

File: src/access/AccessControl.h
Comment (mlepage-google) [reply src/access/AccessControl.h:561]: This PR is going to have to assume persistent storage works. The sooner this PR is merged, the sooner I can work on future things such as "what if it fails half way through" which of course are going to have to be handled.

File: src/app/server/AclStorage.cpp
Comment (tcarmelveilleux) [inline src/app/server/AclStorage.cpp:20]: Please remove this dependency, no loinger used

File: src/app/server/Server.cpp
Comment (tcarmelveilleux) [inline src/app/server/Server.cpp:137]: Let's make AclStorage an interface and inject it in initParams, so it can be overridden if needed. Can be done as a follow-up

File: src/app/server/AclStorage.cpp
Comment (tcarmelveilleux) [inline src/app/server/AclStorage.cpp:444]: Validation of the fabric table and its iteration should probably be done outside, to avoid other maintainers trying to possibly do more, or even own a reference to the fabric table, when it's just used for basic iteration of the list of fabrics.

Suggest adding `AclStorage::InitForFabric(XXX)` 

Also suggest making sure we have an AccessControl call for when a fabric is CREATED to start "empty" in case previous storage errors had left some content that would be unwittingly reused. Can be done in follow-up with a new issue.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/server/AclStorage.cpp
Comment (mlepage-google) [reply src/app/server/AclStorage.cpp]: Done.

File: src/access/AccessControl.h
Comment (mlepage-google) [reply src/access/AccessControl.h:317]: Mentioned in #17912.

File: src/access/AccessControl.h
Comment (mlepage-google) [reply src/access/AccessControl.h:561]: Mentioned in #17912.

File: src/app/server/Server.cpp
Comment (mlepage-google) [reply src/app/server/Server.cpp:137]: Mentioned in #17912.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by mlepage-google** [comment]
Confirm that this work is done now.

**Pull request merged**
